The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5805734372/job/15737300558) [(plain)](https://github.com/rust-lang/rust/commit/b2e95ad62b2f32f13f18feeee6e5799c5b9d9093/checks/15737300558/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
##[endgroup]
Built container sha256:28274c391cc8f7d5f9c48b3e19b3fe4df5df0840d5af7e092c0d03ab7eae1db7
Uploading finished image to https://ci-caches.rust-lang.org/docker/2bcca7776859e4cadc928d5b0475326a46a61958151553d4f5843856969f224dd29cb17625757484fbfc60ad003aadf94c4eed2bc9cdab467e02d68b9cdfc40d

<botocore.awsrequest.AWSRequest object at 0x7f48db855e10>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
[CI_JOB_NAME=x86_64-gnu-llvm-15]
[CI_JOB_NAME=x86_64-gnu-llvm-15]
---
........................................................................................ 13640/15404
........................................................................................ 13728/15404
........................................................................................ 13816/15404
........................................................................................ 13904/15404
......................................................i.....F.F......................... 13992/15404
........................................................................................ 14168/15404
........................................................................................ 14256/15404
........................................................................................ 14344/15404
........................................................................................ 14432/15404
---

---- [ui] tests/ui/for/issue-20605.rs#next stdout ----
diff of stderr:

34 LL |     for item in *things { *item = 0 }
36 
36 
+ error[E0614]: type `<<dyn Iterator<Item = &'a mut u8> as IntoIterator>::IntoIter as Iterator>::Item` cannot be dereferenced
+    |
+    |
+ LL |     for item in *things { *item = 0 }
+ 
+ 
37 error[E0277]: the size for values of type `<<dyn Iterator<Item = &'a mut u8> as IntoIterator>::IntoIter as Iterator>::Item` cannot be known at compilation time
39    |


59    = help: the trait `Sized` is not implemented for `<<dyn Iterator<Item = &'a mut u8> as IntoIterator>::IntoIter as Iterator>::Item`
60 note: required by a bound in `None`
61   --> $SRC_DIR/core/src/option.rs:LL:COL
- 
- error[E0614]: type `<<dyn Iterator<Item = &'a mut u8> as IntoIterator>::IntoIter as Iterator>::Item` cannot be dereferenced
-    |
-    |
- LL |     for item in *things { *item = 0 }
68 
69 error: aborting due to 9 previous errors
70 

---
To only update this specific test, also pass `--test-args for/issue-20605.rs`

error in revision `next`: 1 errors occurred comparing output.
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/for/issue-20605.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--cfg" "next" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/for/issue-20605.next" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/for/issue-20605.next/auxiliary" "-Ztrait-solver=next"
--- stderr -------------------------------
error[E0277]: the trait bound `dyn Iterator<Item = &'a mut u8>: IntoIterator` is not satisfied
##[error]  --> /checkout/tests/ui/for/issue-20605.rs:5:17
   |
   |
LL |     for item in *things { *item = 0 }
   |                 ^^^^^^^ the trait `IntoIterator` is not implemented for `dyn Iterator<Item = &'a mut u8>`

error[E0277]: the size for values of type `<dyn Iterator<Item = &'a mut u8> as IntoIterator>::IntoIter` cannot be known at compilation time
   |
   |
LL |     for item in *things { *item = 0 }
   |
   |
   = help: the trait `Sized` is not implemented for `<dyn Iterator<Item = &'a mut u8> as IntoIterator>::IntoIter`
   = help: unsized locals are gated as an unstable feature


error: the type `<dyn Iterator<Item = &'a mut u8> as IntoIterator>::IntoIter` is not well-formed
   |
   |
LL |     for item in *things { *item = 0 }


error[E0277]: `<dyn Iterator<Item = &'a mut u8> as IntoIterator>::IntoIter` is not an iterator
   |
   |
LL |     for item in *things { *item = 0 }
   |                 ^^^^^^^ `<dyn Iterator<Item = &'a mut u8> as IntoIterator>::IntoIter` is not an iterator
   |
   = help: the trait `Iterator` is not implemented for `<dyn Iterator<Item = &'a mut u8> as IntoIterator>::IntoIter`

error: the type `&mut <dyn Iterator<Item = &'a mut u8> as IntoIterator>::IntoIter` is not well-formed
   |
   |
LL |     for item in *things { *item = 0 }


error[E0614]: type `<<dyn Iterator<Item = &'a mut u8> as IntoIterator>::IntoIter as Iterator>::Item` cannot be dereferenced
   |
   |
LL |     for item in *things { *item = 0 }


error[E0277]: the size for values of type `<<dyn Iterator<Item = &'a mut u8> as IntoIterator>::IntoIter as Iterator>::Item` cannot be known at compilation time
   |
   |
LL |     for item in *things { *item = 0 }
   |
   |
   = help: the trait `Sized` is not implemented for `<<dyn Iterator<Item = &'a mut u8> as IntoIterator>::IntoIter as Iterator>::Item`
   = help: unsized locals are gated as an unstable feature

Build completed unsuccessfully in 0:10:51
Build completed unsuccessfully in 0:10:51
error: the type `Option<<<dyn Iterator<Item = &'a mut u8> as IntoIterator>::IntoIter as Iterator>::Item>` is not well-formed
   |
   |
LL |     for item in *things { *item = 0 }


error[E0277]: the size for values of type `<<dyn Iterator<Item = &'a mut u8> as IntoIterator>::IntoIter as Iterator>::Item` cannot be known at compilation time
   |
   |
LL |     for item in *things { *item = 0 }
   |
   |
   = help: the trait `Sized` is not implemented for `<<dyn Iterator<Item = &'a mut u8> as IntoIterator>::IntoIter as Iterator>::Item`
note: required by a bound in `None`
  --> /rustc/FAKE_PREFIX/library/core/src/option.rs:567:5
error: aborting due to 9 previous errors

Some errors have detailed explanations: E0277, E0614.
For more information about an error, try `rustc --explain E0277`.
For more information about an error, try `rustc --explain E0277`.
------------------------------------------


---- [ui] tests/ui/lazy-type-alias/coerce-behind-lazy.rs#next stdout ----

error in revision `next`: test compilation failed although it shouldn't!
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lazy-type-alias/coerce-behind-lazy.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--cfg" "next" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lazy-type-alias/coerce-behind-lazy.next" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lazy-type-alias/coerce-behind-lazy.next/auxiliary" "-Ztrait-solver=next"
--- stderr -------------------------------
warning: the feature `lazy_type_alias` is incomplete and may not be safe to use and/or cause compiler crashes
##[warning]  --> /checkout/tests/ui/lazy-type-alias/coerce-behind-lazy.rs:5:12
   |
---

error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/lazy-type-alias/coerce-behind-lazy.rs:13:5
   |
LL | fn test() -> Coerce {
   |              ------ expected `Coerce` because of return type
LL |     Box::new(1)
   |     ^^^^^^^^^^^- help: try using a conversion method: `.to_string()`
   |     types differ
   |
   = note: expected type alias `Coerce`
   = note: expected type alias `Coerce`
                  found struct `Box<{integer}>`
error: aborting due to previous error; 1 warning emitted

For more information about this error, try `rustc --explain E0308`.
------------------------------------------
------------------------------------------


---- [ui] tests/ui/traits/new-solver/more-object-bound.rs stdout ----
diff of stderr:

1 error[E0277]: the trait bound `dyn Trait<A = A, B = B>: Trait` is not satisfied
+   --> $DIR/more-object-bound.rs:12:17
3    |
3    |
4 LL |     foo::<A, B, dyn Trait<A = A, B = B>>(x)
-    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Trait` is not implemented for `dyn Trait<A = A, B = B>`
+    |                 ^^^^^^^^^^^^^^^^^^^^^^^ the trait `Trait` is not implemented for `dyn Trait<A = A, B = B>`
7 note: required by a bound in `foo`
8   --> $DIR/more-object-bound.rs:18:8



The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/traits/new-solver/more-object-bound/more-object-bound.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args traits/new-solver/more-object-bound.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/traits/new-solver/more-object-bound.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/traits/new-solver/more-object-bound" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/traits/new-solver/more-object-bound/auxiliary" "-Ztrait-solver=next"
--- stderr -------------------------------
--- stderr -------------------------------
error[E0277]: the trait bound `dyn Trait<A = A, B = B>: Trait` is not satisfied
   |
   |
LL |     foo::<A, B, dyn Trait<A = A, B = B>>(x)
   |                 ^^^^^^^^^^^^^^^^^^^^^^^ the trait `Trait` is not implemented for `dyn Trait<A = A, B = B>`
note: required by a bound in `foo`
  --> /checkout/tests/ui/traits/new-solver/more-object-bound.rs:18:8
   |
   |
LL | fn foo<A, B, T: ?Sized>(x: T::A) -> B
LL | where
LL | where
LL |     T: Trait<B = B>,
   |        ^^^^^^^^^^^^ required by this bound in `foo`
help: consider introducing a `where` clause, but there might be an alternative better way to express this requirement
   |
LL | fn transmute<A, B>(x: A) -> B where dyn Trait<A = A, B = B>: Trait {

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
For more information about this error, try `rustc --explain E0277`.
------------------------------------------


---- [ui] tests/ui/traits/new-solver/lazy-nested-obligations-2.rs stdout ----

error: test compilation failed although it shouldn't!
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/traits/new-solver/lazy-nested-obligations-2.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/traits/new-solver/lazy-nested-obligations-2" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/traits/new-solver/lazy-nested-obligations-2/auxiliary" "-Ztrait-solver=next"
--- stderr -------------------------------
error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/traits/new-solver/lazy-nested-obligations-2.rs:15:23
   |
   |
LL |     let _: V<i32> = V(f);
   |                     - ^ types differ
   |                     arguments to this struct are incorrect
   |
   |
   = note: expected associated type `<i32 as With>::F`
                      found fn item `for<'a> fn(&'a str) {f}`
   = help: consider constraining the associated type `<i32 as With>::F` to `for<'a> fn(&'a str) {f}` or calling a method that returns `<i32 as With>::F`
note: tuple struct defined here
  --> /checkout/tests/ui/traits/new-solver/lazy-nested-obligations-2.rs:16:16
   |
   |
LL |     pub struct V<T: With>(<T as With>::F);

error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/traits/new-solver/lazy-nested-obligations-2.rs:21:30
   |
   |
LL |     let _: E3<i32> = E3::Var(f);
   |                      ------- ^ types differ
   |                      arguments to this enum variant are incorrect
   |
   |
   = note: expected associated type `<i32 as With>::F`
                      found fn item `for<'a> fn(&'a str) {f}`
   = help: consider constraining the associated type `<i32 as With>::F` to `for<'a> fn(&'a str) {f}` or calling a method that returns `<i32 as With>::F`
note: tuple variant defined here
  --> /checkout/tests/ui/traits/new-solver/lazy-nested-obligations-2.rs:19:9
   |
   |
LL |         Var(<T as With>::F),

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0308`.
```

</details>

--- a/compiler/rustc_hir_typeck/src/coercion.rs
+++ b/compiler/rustc_hir_typeck/src/coercion.rs
@@ -1007,15 +1007,17 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         &self,
         expr: &hir::Expr<'_>,
         expr_ty: Ty<'tcx>,
-        target: Ty<'tcx>,
+        mut target: Ty<'tcx>,
         allow_two_phase: AllowTwoPhase,
         cause: Option<ObligationCause<'tcx>>,
     ) -> RelateResult<'tcx, Ty<'tcx>> {
         let source = self.try_structurally_resolve_type(expr.span, expr_ty);
-        let target = self.try_structurally_resolve_type(
-            cause.as_ref().map_or(expr.span, |cause| cause.span),
-            target,
-        );
+        if self.next_trait_solver() {
+            target = self.try_structurally_resolve_type(
+                cause.as_ref().map_or(expr.span, |cause| cause.span),
+                target,
+            );
+        }
         debug!("coercion::try({:?}: {:?} -> {:?})", expr, source, target);
 
         let cause =

